### PR TITLE
Feature/improve godata

### PR DESCRIPTION
**Pull request by ebuchman**
Openning PR for review

### CODE REVIEW

File: encode_test.go [hunk 1]
```diff
@@ -95,17 +115,17 @@ func toFromJSON(t *testing.T, in interface{}, recvr interface{}) {
 
 func TestNilEncodings(t *testing.T) {
 	// make sure sigs are okay with nil
-	a, b := SignatureS{}, SignatureS{}
+	a, b := Signature{}, Signature{}
```
Comment (ebuchman) [inline encode_test.go]: `var a, b Signature`

**Review by ebuchman — COMMENTED** [review]
Looks good. I might stress a little more that the XXXInner interfaces are not to be used ....

File: embed_test.go [hunk 2]
```diff
@@ -0,0 +1,125 @@
...
+	// assert := assert.New(t)
+	// require := require.New(t)
+
+	cases := []struct {
+		in, out  Greeter
+		expected string
+	}{
+		{PubName{Foo{"pub-foo"}}, &PubName{}, "Foo: pub-foo"},
+		{PubName{Bar{7}}, &PubName{}, "Bar #7"},
+		// Note these fail - if you can figure a solution here, I'll buy you a beer :)
```
Comment (ebuchman) [inline embed_test.go]: don't these fail because the interface is private and hence can't be seen by wire ?

File: encode_test.go (hunk 1 above)
Comment (ethanfrey) [reply encode_test.go]: Yes, that is more idiomatic. Thanks

How to stress do not use except by changing the names? I'm open to suggestions

Comment (ebuchman) [reply encode_test.go]: Comments are fine:

`DO NOT USE this interface. It is public by necessity but should never be used directly outside of this package. Only use the PubKey, never the PubKeyInner`

File: embed_test.go (hunk 2 above)
Comment (ethanfrey) [reply embed_test.go]: Exactly. I wrote the tests to verify the behavior before coverting all code.

Comment (ebuchman) [reply embed_test.go]: Ok. So there's no real beer offering here and we can remove these ? Or do you want to assert in a separate test that they'll fail?

Comment (ethanfrey) [reply embed_test.go]: I'd love a trick to make them private without breaking go-wire. Thus, if you can fix these tests, congrats!

Comment (ebuchman) [reply embed_test.go]: Wouldn't that be breaking the type system? There should be no way to get an external package to read private stuff - though I'm not sure what happens if the field is public but the type is private

Comment (ethanfrey) [reply embed_test.go]: I don't mean breaking the type system. But I got to use private fields in json by implementing custom Marshalers. I guess there is no equivalent in go- wire?
Like the fields are private, but there are public methods that modify them.

File: encode_test.go (hunk 1 above)
Comment (ethanfrey) [reply encode_test.go]: Awesome. Will do tomorrow

**Comment by ethanfrey** [comment]
I think I fixed the issues.

The failing test is due to it requiring the develop branch of go-data (which has the breaking change of renaming RegisterInterface to RegisterImplementation).  And go-crypto has no glide file or `make get_vendor_deps`.  Should I add that???

**Comment by ethanfrey** [comment]
The go-wire struct parsing seems to be dealt with [here](https://github.com/tendermint/go-wire/blob/master/reflect.go#L314-L318) and [here](https://github.com/tendermint/go-wire/blob/master/reflect.go#L524-L528)

They make use of `reflect.Value.Field` and the `TypeInfo` list that was created [here](https://github.com/tendermint/go-wire/blob/master/reflect.go#L148-L182)

Assuming the go-wire code were modified, one could add a simple interface to allow custom un/marshaling of go-wire binary, like:

```Go
type Reflector interface {
  NumFields() int
  Field(int) interface{}  // returning a pointer to the value at that location
}
```

Probably totally out of scope and unnecessary, but I write it here for what it is worth.  If you think it is useful, I can make an issue in go-wire...

**Pull request merged**